pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,244 @@ var (
 	}
 )
 
-// Requirements is a decorated alias type for []v1.NodeSelectorRequirements
-type Requirements []v1.NodeSelectorRequirement
-
-func (r Requirements) Zones() sets.String {
-	return r.Requirement(v1.LabelTopologyZone)
-}
-
-func (r Requirements) InstanceTypes() sets.String {
-	return r.Requirement(v1.LabelInstanceTypeStable)
+type Requirements struct {
+	// Requirements are layered with Labels and applied to every node.
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	// Ground truth for record keeping
+	Records map[string][]v1.NodeSelectorRequirement `json:"-"`
```

Comment:
It's not clear to me why we need this instead of just Requirements?

Reply:
In this commit `Requirements []v1.NodeSelectorRequirement` is not used at all other than API codegen.
I chose to use `Records map[string][]v1.NodeSelectorRequirement` to simplify the codes by geting ride of all indexing for loops. It is also faster.